packages/dd-trace/test/datastreams/data_streams_checkpointer.spec.js

```diff
@@ -0,0 +1,71 @@
+require('../setup/tap')
+
+const agent = require('../plugins/agent')
+
+const expectedProducerHash = '13182885521735152072'
+const expectedConsumerHash = '5980058680018671020'
+const DSM_CONTEXT_HEADER = 'dd-pathway-ctx-base64'
+
+describe('data streams checkpointer manual api', () => {
+  let tracer
+
+  before(() => {
+    process.env.DD_DATA_STREAMS_ENABLED = 'true'
+    tracer = require('../..').init()
+    agent.load(null, { dsmEnabled: true })
+  })
+
+  after(() => {
+    return agent.close({ ritmReset: false })
+  })
+
+  it('should set a checkpoint when calling setProduceCheckpoint', function (done) {
+    const expectedEdgeTags = ['direction:out', 'manual_checkpoint:true', 'topic:test-queue', 'type:testProduce']
+
+    agent.expectPipelineStats(dsmStats => {
+      let statsPointsReceived = 0
+      // we should have 1 dsm stats points
+      dsmStats.forEach((timeStatsBucket) => {
+        if (timeStatsBucket && timeStatsBucket.Stats) {
+          timeStatsBucket.Stats.forEach((statsBuckets) => {
+            statsPointsReceived += statsBuckets.Stats.length
+          })
+        }
+      })
```

Comment:
It's best to avoid `forEach(...)` as it's generally considered legacy JS at this point and is not well supported by the optimizer. You could do something like:

```suggestion
      let statsPointsReceived = 0
      // we should have 1 dsm stats points
      for (const timeStatsBucket of dsmStats) {
        if (timeStatsBucket && timeStatsBucket.Stats) {
          for (const statsBucket of timeStatsBucket) {
            statsPointsReceived += statsBuckets.Stats.length
          })
        }
      })
```

If you want to take a more functional approach you could do something more like:

```suggestion
      // we should have 1 dsm stats points
      const statsPointsReceived = dsmStats
        .filter((bucket) => bucket && bucket.Stats)
        .flatMap((bucket) => bucket.Stats)
        .map((stats) => stats.Stats.length)
        .reduce((m, v) => m + v, 0)
```

Reply:
Removed all uses of `.forEach`